src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/ProjectServices/CpsProjectSystemReferenceReader.cs

```diff
@@ -0,0 +1,114 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Globalization;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft;
+using Microsoft.VisualStudio.ProjectSystem;
+using Microsoft.VisualStudio.ProjectSystem.Properties;
+using Microsoft.VisualStudio.ProjectSystem.References;
+using Microsoft.VisualStudio.Threading;
+using NuGet.Frameworks;
+using NuGet.LibraryModel;
+using NuGet.ProjectManagement;
+using NuGet.ProjectModel;
+using NuGet.VisualStudio;
+
+namespace NuGet.PackageManagement.VisualStudio
+{
+    /// <summary>
+    /// Reference reader implementation for the core project system in the integrated development environment (IDE).
+    /// </summary>
+    internal class CpsProjectSystemReferenceReader
+        : IProjectSystemReferencesReader
+    {
+        private readonly IVsProjectAdapter _vsProjectAdapter;
+        private readonly IVsProjectThreadingService _threadingService;
+        private readonly AsyncLazy<ConfiguredProject> _configuredProject;
+
+        public CpsProjectSystemReferenceReader(
+            IVsProjectAdapter vsProjectAdapter,
+            IVsProjectThreadingService threadingService)
+        {
+            Assumes.Present(vsProjectAdapter);
+            Assumes.Present(threadingService);
+
+            _vsProjectAdapter = vsProjectAdapter;
+            _threadingService = threadingService;
+            _configuredProject = new AsyncLazy<ConfiguredProject>(async () =>
+            {
+                await _threadingService.JoinableTaskFactory.SwitchToMainThreadAsync();
+
+                var context = _vsProjectAdapter.Project as IVsBrowseObjectContext;
```

Comment:
I had a look at this, and for C++, this is just going to pick whatever the first configuration is. For other CPS projects, this is always going to return null.

Reply:
I think instead you want to capture the `UnconfiguredProject` and remember that in this AsyncLazy.

Then when you go retrieve the references, call `UnconfiguredProject.GetSuggestedConfiguredProjectAsync()` and handle null;

Reply:
Sounds good. 

I applied that feedback. Basic tests suggest everything works as expected.